### ISSUE TITLE
How to add validation loss calculation during training?

### ISSUE THREAD

**Issue by 08D20088** [2U1/Gemma3-Finetune#9]
@2U1 

Hi, thank you for the great work on Gemma3-Finetune.
Currently, the training script only reports training loss.
It would be very helpful if the code could also evaluate and log the validation loss during training (e.g., at the end of each epoch).

This would make it easier to monitor overfitting and model performance while fine-tuning.
I’d like to calculate validation loss during training (e.g., at the end of each epoch), but I’m not sure where in the code I should modify.

Where should I insert validation evaluation?

**Comment by 2U1** [comment]
You could add 
```
def prediction_step(self, model, inputs, prediction_loss_only, ignore_keys=None):
        labels = inputs.get("labels") if "labels" in inputs else None

        with torch.no_grad():
            outputs = model(**inputs)
            loss = outputs.loss if hasattr(outputs, "loss") else None
            logits = outputs.logits if hasattr(outputs, "logits") else None

        if prediction_loss_only:
            return (loss, None, None)
        return (loss, logits, labels)
``` 
This to the sft_trainer.

Also you should fix the code to add the eval dataset set and pass to the trainer. Thats not a really hard part.

**Comment by 08D20088** [comment]
Thank you for your quick response.
I added the `prediction_step` function you provided to the `GemmaSFTTrainer` class in `sft_trainer.py`, and made some additional modifications to the code.

After that, I ran `finetune.sh` and got the following outputs:

```bash
{'loss': 20.3067, 'grad_norm': 2048.30895248638, 'learning_rate': 0.0, 'epoch': 0.01}
{'eval_loss': 1.0087, 'eval_runtime': 76.4788, 'eval_samples_per_second': 13.076, 'eval_steps_per_second': 0.235, 'epoch': 0.07}
```

It seems that the implementation is working, but since I am a beginner, I’m worried I might have overlooked something.
Here are the changes I made—could you please let me know if there are any other modifications I should apply?

---

### Changes

#### Added the following arguments to `deepspeed` in `finetune.sh`

```bash
--eval_data_path $eval_data_json
--eval_image_folder $eval_image_folder
--eval_strategy steps
--eval_steps 1
```

#### Added the following arguments to `SupervisedDataset` class in `sft_dataset.py`

```python
    def __init__(
        self,
        data_path: str | list,
        processor: transformers.ProcessorMixin,
        data_args: DataArguments,
        padding=True,
        image_folder: str = None,
    ):
```

#### Modified the `make_supervised_data_module` function in `sft_dataset.py`

```python
def make_supervised_data_module(processor, data_args):
    """Make dataset and collator for supervised fine-tuning."""
    sft_dataset = SupervisedDataset(
        data_path=data_args.data_path, processor=processor, data_args=data_args
    )

    eval_dataset = None
    if data_args.eval_data_path:
        eval_dataset = SupervisedDataset(
            data_path=data_args.eval_data_path, 
            processor=processor, 
            data_args=data_args,
            image_folder=data_args.eval_image_folder
        )

    data_collator = DataCollatorForSupervisedDataset(
        pad_token_id=processor.tokenizer.pad_token_id
    )

    return dict(
        train_dataset=sft_dataset,
        eval_dataset=eval_dataset,
        data_collator=data_collator
    )
```

#### Added `eval_data_path` and `eval_image_folder` to the `DataArguments` class in `params.py`

```python
class DataArguments:
    data_path: str = field(
        default=None, metadata={"help": "Path to the training data."}
    )
    eval_data_path: str = field(
        default=None, metadata={"help": "Path to the evaluation data."}
    )
    lazy_preprocess: bool = False
    image_folder: Optional[str] = field(default=None)
    eval_image_folder: Optional[str] = field(default=None)
    max_num_frames: int = 10
```

**Comment by 2U1** [comment]
Yes you are doing right.
You could measure some other metric by adding compute_metrics other than loss/perplexity.

**Comment by 08D20088** [comment]
Thank you so much.
